reframe/core/decorators.py

```diff
@@ -27,33 +27,33 @@
 from reframe.utility.versioning import VersionValidator
 
 
-def _register_test(cls, args=None):
+def _register_test(cls, inst_args=None, *args, **kwargs):
     '''Register the test.
 
     Register the test with _rfm_use_params=True. This additional argument flags
     this case to consume the parameter space. Otherwise, the regression test
     parameters would simply be initialized to None.
     '''
-    def _instantiate(cls, args):
-        if isinstance(args, collections.abc.Sequence):
-            return cls(*args, _rfm_use_params=True)
-        elif isinstance(args, collections.abc.Mapping):
-            args['_rfm_use_params'] = True
-            return cls(**args)
-        elif args is None:
-            return cls(_rfm_use_params=True)
-
-    def _instantiate_all():
+    def _instantiate(cls, inst_args, *args, **kwargs):
+        kwargs['_rfm_use_params'] = True
+        if isinstance(inst_args, collections.abc.Sequence):
+            args += inst_args
+        elif isinstance(inst_args, collections.abc.Mapping):
+            kwargs.update(args)
+
+        return cls(*args, **kwargs)
+
+    def _instantiate_all(*args, **kwargs):
```

Comment:
Rather than setting this `_instantiate_all` function as the module's `_rfm_gettests`, I think it would make more sense to make this test registry a class such that the loader can access each of the registered classes in it. Then the `instantiate_all` method can be a public member function that the loader calls.

Also, that would solve the whole `inst_args`, `*args`, `**kwargs` issue here, which is rather complex otherwise.